daffodil-lib/src/main/scala/org/apache/daffodil/lib/xml/DaffodilXMLLoader.scala

```diff
@@ -250,7 +262,11 @@ class DFDLCatalogResolver private ()
         // use the systemId (which comes from the schemaLocation attribute)
         // and the classpath.
         val baseURI = if (baseURIString == null) None else Some(new URI(baseURIString))
-        val optURI = Misc.getResourceRelativeOption(sysId, baseURI)
+        // try it as a direct classpath resolution first, and if that fails,
+        // try removing all the upward path steps (if any).
+        val optURI = Misc.getResourceRelativeOption(sysId, baseURI).orElse(
+          Misc.getResourceRelativeOption(removeInitialUpwardPathSteps(sysId), baseURI)
```

Comment:
Well, the corner case you suggest does and could happen.

But not if schemas use package-like names for directories, or very unique file names. 

It certainly is possible to get the number of "../" wrong, but it's also possible to cut/paste the wrong root-path component name and then forget to edit it. 

One nice thing is for the relative paths within a schema, the IDE highlights them if they're wrong.  Not inter-schema however. 

I just tested all this with a few schemas. I used Janap128 and OTH-Gold (which are Owl's schemas). I changed all the schemaLocation paths that were relative to use relative up-steps, and the classpath/inter-component ones to have up-steps to the root.  

All worked first try, once I put this same resolver change into the Owl xsat2 library. 

I suggest this: we shouldn't let the perfect be the enemy of the good here. This change will really help out people trying to use DFDL schemas to validate XML data using, well, not-daffodil. Ex: C-based tooling. 

If debugging include/import schemaLocations becomes problematic, then we can make changes to getResourceRelativeOption along the lines of what you are suggesting. We're writing our own class path search in that case but that would be ok because we could get much better diagnostics.

Reply:
The more I think about it, the more it feels wrong to just throw away parts of an include path. If we're five directories deep and fail to resolve `../foo.dfdl.xsd`, we probably shouldn't look for `/foo.dfdl.xsd` on the classpath, it's very unlikely that's what the user intended and could easily find the wrong thing.

How about this alternative approach, which doesn't handle too many up-steps but I think is maybe less complex and maintains directory namespaceing: We generate a relative URL the same way we do now. If that fails to resolve and if it's a jar URI then we extract the new path from it and look that up on the classpath.

For example, if the base URI is `jar:file://path/to/foo.jar!/com/example/xsd/main.dfdl.xsd` and that includes `../common.dfdl.xsd` then we first look for `jar:file://path/to/foo.jar!/com/example/common.dfdl.xsd`. If that fails to resolve, then we look up `/com/example/common.dfdl.xsd` on the classpath. This way we don't lose our namespace information. 

I think the code would be fairly straightforward:

```scala
def getResourceRelativeOption(...) : Option[URI] = {
{
  ...
  val completeURL =  new URL(contextURL, resName)
  ...
  val res = ...
  // new stuff here
  if (res.isEmpty && completeURL.getProtocol == "jar") {  
    val newPath = completeURL.openConnection().asInstanceOf[JarURLConnection].getEntryName
    Misc.getResourceOption(newPath)
  } else {
    res
  }
}
```

This doesn't detect too-many up-steps, but it doesn't throwaway relative path information so with namespaced directories makes it unlikely to find the wrong thing. And if relative steps go all the way up to the root of the jar (or beyond) then it essentially behaves the same as this PR and looks on the classpath with all relative steps gone.

Reply:
I think this won't work because a classpath can contain more than just jars. It can contain an arbitrary sequence of directories. So the URI isn't necessarily a jar URI.

Reply:
How about this. (1) The schemaLocation gets combined with the current file URI to form a combined URI. If that then resolves, great. 
If it doesn't work, then by cases:

(2) If the current file URI is a jar URI, then detect the "too many up steps" error. If the up-steps take you to the root of the jar, then the schemaLocation without the up-steps is resolved against the class path. 
(3) If the up-steps do not take you to the root of the jar, then fail - should have worked at step (1)

(4) If the current file URI is NOT a jar, then see if the directories path steps immediately above all the ../ moves are named "src/*/resources" where * is main or test. If so, then heuristically we are at the root, so try the schemaLocation without the up-steps against the class path. If not fail - should have worked at step (1). 
If the up-steps take you to within or above the src/main/resources or src/test/resources (or any other heuristic that names the root dir of a resource tree) then you have too many up steps error. 

I think other similar heuristics would look to see if the up-steps take you to a resource_managed directory.

Reply:
Additionally to 4, we may also want to consider the flattened directory layout, where there is no {main,test}/resources and everything is in just in src/ or test/. 

Though, I wonder there is a potential issue with number 4, where things might work fine when running in the project repo because it sees src/*/resources parent dirs and switches to the classpath lookup, but then fails if copied somewhere else (e.g. some product) where those src/*/resources directories don't exist and so it can't know to use the classpath?

I'm not sure what could be done about this though...